packages/contracts/contracts/router/modules/exchanges/RaribleModule.sol

```diff
@@ -0,0 +1,288 @@
+        bytes calldata signatureRight,
+        OfferParams calldata params,
+        Fee[] calldata fees
+    ) external nonReentrant {
+        (address token, uint tokenId) = abi.decode(orderLeft.makeAsset.assetType.data, (address, uint256));
+
+        IERC1155 collection = IERC1155(address(token));
+
+        // Approve the transfer manager if needed
+        _approveERC1155IfNeeded(collection, TRANSFER_MANAGER);
+
+        // Execute the fill
+        _sell(
+            orderLeft,
+            signatureLeft,
+            orderRight,
+            signatureRight,
+            params.fillTo,
+            params.revertIfIncomplete,
+            fees
+        );
+
+        // Refund any ERC1155 leftover
+        _sendAllERC1155(params.refundTo, collection, tokenId);
+    }
+
+    // --- ERC721 / ERC1155 hooks ---
+
+    // Single token offer acceptance can be done approval-less by using the
+    // standard `safeTransferFrom` method together with specifying data for
+    // further contract calls. An example:
+    // `safeTransferFrom(
+    //      0xWALLET,
+    //      0xMODULE,
+    //      TOKEN_ID,
+    //      0xABI_ENCODED_ROUTER_EXECUTION_CALLDATA_FOR_OFFER_ACCEPTANCE
+    // )`
+
+    function onERC721Received(
+        address, // operator,
+        address, // from
+        uint256, // tokenId,
+        bytes calldata data
+    ) external returns (bytes4) {
+        if (data.length > 0) {
+            _makeCall(router, data, 0);
+        }
+
+        return this.onERC721Received.selector;
+    }
+
+    function onERC1155Received(
+        address, // operator
+        address, // from
+        uint256, // tokenId
+        uint256, // amount
+        bytes calldata data
+    ) external returns (bytes4) {
+        if (data.length > 0) {
+            _makeCall(router, data, 0);
+        }
+
+        return this.onERC1155Received.selector;
+    }
+
+    // --- Internal ---
+
+    function _buy(
+        IRarible.Order calldata orderLeft,
+        bytes calldata signatureLeft,
+        IRarible.Order calldata orderRight,
+        bytes calldata signatureRight,
+        address receiver,
+        bool revertIfIncomplete,
+        uint256 value
+    ) internal {
+        // Execute the fill
+        try
+            EXCHANGE.matchOrders{value: value}(orderLeft, signatureLeft, orderRight, signatureRight)
+        {
+            (address token, uint tokenId) = abi.decode(orderLeft.makeAsset.assetType.data, (address, uint256));
+
+            IERC165 collection = IERC165(address(token));
+
+            // Forward any token to the specified receiver
+            bool isERC721 = collection.supportsInterface(ERC721_INTERFACE);
+            if (isERC721) {
+                IERC721(address(collection)).safeTransferFrom(
+                    address(this),
+                    receiver,
+                    tokenId
+                );
+            } else {
+                bool isERC1155 = collection.supportsInterface(
```

Comment:
We can skip `isERC1155` and assume that the `else` branch will always be the `ERC1155` case.